### PR TITLE
feat: M3 batch 2 — overlay promote, overlay UI, changeset

### PR DESCRIPTION
**Pull request by laofahai**
## Summary
- **Overlay Promote** (`linch overlay promote`): graduate overlay fields to code-level entity definitions, with dry-run, backup, and diff support
- **Overlay UI**: AutoForm + AutoList render overlay fields with visual badges, using `useOverlayFields` hook
- **Changeset**: initial 0.1.0 release changeset
- **CI fix**: increase watcher staleness test timeout for stability

## Changes
- 21 files changed, +1290 lines
- New CLI command: `overlay promote`
- New UI components: `OverlayFieldBadge`, `useOverlayFields` hook
- New core module: `packages/core/src/overlay/promote.ts`
- 284-line test suite for overlay promote

## Test plan
- [x] `bun test` — 3804 pass (49 pre-existing DB integration failures)
- [x] `bun run typecheck` — clean
- [x] `bun run lint` — clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Overlay (custom) fields can be added as extra columns in list views and shown in forms.
  * Visual badge marks overlay/custom fields in forms and lists.
* **Improvements**
  * Overlay columns support sorting for string/number/date types.
  * Form submissions now preserve existing custom field data; overlay labels better align with browser locale.
* **Tests**
  * Improved test resilience for edge cases.

### CODE REVIEW

File: addons/adapter-ui/cap-adapter-ui/src/components/auto-form/auto-form.tsx [hunk 1]
```diff
@@ -425,6 +441,21 @@ export function AutoForm({
...
+        const key = overlayFieldKey(overlay.fieldName);
+        if (key in submitData) {
+          extensions[overlay.fieldName] = submitData[key];
+          delete submitData[key];
+        }
+      }
+      if (Object.keys(extensions).length > 0) {
+        submitData._extensions = extensions;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (laofahai) [reply addons/adapter-ui/cap-adapter-ui/src/components/auto-form/auto-form.tsx:482]: Fixed. Now merging existing _extensions data via spread before overlaying new values.

File: addons/adapter-ui/cap-adapter-ui/src/components/auto-form/auto-form.tsx [hunk 2]
```diff
@@ -787,6 +861,101 @@ function getDefaultForType(fieldDef: FieldDefinition): unknown {
...
+  if (!overlay.config.label) return overlay.fieldName;
+  // Try browser language first, then English fallback
+  const lang = typeof navigator !== "undefined" ? navigator.language.replace("-", "-") : "en";
+  return (
+    overlay.config.label[lang] ??
+    overlay.config.label.en ??
+    Object.values(overlay.config.label)[0] ??
+    overlay.fieldName
+  );
+}
```
[the remark this answers is not in the document]
Comment (laofahai) [reply addons/adapter-ui/cap-adapter-ui/src/components/auto-form/auto-form.tsx:1033]: Fixed. Removed the no-op .replace, now using navigator.language directly.

File: packages/cli/src/commands/overlay.ts [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+      for (const overlay of targets) {
+        const plan = generatePromotionPlan(overlay);
+
+        // Write migration SQL file
+        const migFile = join(
+          migDir,
+          `promote_${overlay.entityName}_${overlay.fieldName}.sql`,
+        );
+        writeFileSync(migFile, plan.migrationSql, "utf-8");
```
[the remark this answers is not in the document]
Comment (laofahai) [reply packages/cli/src/commands/overlay.ts:201]: Fixed. Added sanitizeForFilename() helper — only allows [a-zA-Z0-9_-], replaces everything else with underscore.

File: packages/core/__tests__/watcher-engine.test.ts [hunk 4]
```diff
@@ -879,8 +879,8 @@ describe("WatcherEngine — staleness triggers", () => {
 
     engine.start();
 
-    // Wait for the interval to fire
-    await new Promise((resolve) => setTimeout(resolve, 100));
+    // Wait for the interval to fire (generous timeout for CI)
+    await new Promise((resolve) => setTimeout(resolve, 300));
```
[the remark this answers is not in the document]
Comment (laofahai) [reply packages/core/__tests__/watcher-engine.test.ts]: Fixed. Reduced wait from 300ms to 80ms and moved engine.stop() immediately after await. All 42 watcher tests pass.

File: packages/core/src/overlay/promote.ts [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    );
+  }
+
+  lines.push(
+    "",
+    `-- Step 3: Clean up _extensions (remove promoted key)`,
+    `UPDATE "${tableName}"`,
+    `SET _extensions = _extensions - '${fieldName}'`,
+    `WHERE _extensions ? '${fieldName}';`,
+  );
```
[the remark this answers is not in the document]
Comment (laofahai) [reply packages/core/src/overlay/promote.ts]: Fixed. Added quoteIdentifier() and quoteLiteral() helpers. All SQL interpolations now use these for escaping.

File: packages/core/src/overlay/promote.ts [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+      `SET "${fieldName}" = _extensions->'${fieldName}'`,
+      `WHERE _extensions ? '${fieldName}';`,
+    );
+  } else {
+    lines.push(
+      `UPDATE "${tableName}"`,
+      `SET "${fieldName}" = (_extensions->>'${fieldName}')::${castType}`,
+      `WHERE _extensions ? '${fieldName}';`,
+    );
+  }
```
[the remark this answers is not in the document]
Comment (laofahai) [reply packages/core/src/overlay/promote.ts:182]: Fixed. Required field backfill now uses COALESCE with getDefaultExpression fallback.

File: packages/core/src/overlay/promote.ts [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+      return "''";
+    case "number":
+      return "0";
+    case "boolean":
+      return "FALSE";
+    case "json":
+      return "'null'::jsonb";
+    default:
+      return null;
+  }
```
[the remark this answers is not in the document]
Comment (laofahai) [reply packages/core/src/overlay/promote.ts:238]: Fixed. Added case date returning epoch::timestamptz in getDefaultExpression.

File: packages/core/src/overlay/promote.ts [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+  if (fieldType === "string" && config.maxLength !== undefined) {
+    props.push(`maxLength: ${config.maxLength}`);
+  }
+
+  // Enum options
+  if (fieldType === "enum" && config.enumValues?.length) {
+    props.push(`options: ${JSON.stringify(config.enumValues)}`);
+  }
+
+  return `${fieldName}: { ${props.join(", ")} }`;
```
[the remark this answers is not in the document]
Comment (laofahai) [reply packages/core/src/overlay/promote.ts]: Fixed. Field keys with special characters now use JSON.stringify. Safe identifiers remain unquoted.

File: packages/core/src/overlay/promote.ts [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    }
+    if (config.max !== undefined) {
+      props.push(`max: ${config.max}`);
+    }
+  }
+
+  // String constraints
+  if (fieldType === "string" && config.maxLength !== undefined) {
+    props.push(`maxLength: ${config.maxLength}`);
+  }
```
[the remark this answers is not in the document]
Comment (laofahai) [reply packages/core/src/overlay/promote.ts:104]: Verified — OverlayFieldConfig correctly uses max for numbers and maxLength for strings as separate properties, matching the rest of the codebase. No change needed.

File: addons/adapter-ui/cap-adapter-ui/src/components/auto-form/auto-form.tsx [hunk 10]
```diff
@@ -63,12 +65,16 @@ export function AutoForm({
...
   templates,
+  overlayFields,
   formId: customFormId,
 }: AutoFormProps) {
   const { t } = useTranslation();
   const formId = customFormId ?? "auto-form";
   const zodSchema = useMemo(() => generateZodSchema(schema), [schema]);
 
+  // Convert overlay fields to FieldDefinition map for rendering via widget registry
+  const overlayFieldDefs = useMemo(() => buildOverlayFieldDefs(overlayFields), [overlayFields]);
```
[the remark this answers is not in the document]
Comment (laofahai) [reply addons/adapter-ui/cap-adapter-ui/src/components/auto-form/auto-form.tsx:87]: Fixed. Added `buildOverlayZodField()` helper that maps overlay field config (required, min, max, maxLength, enumValues) to Zod schemas. The base schema is now extended with overlay field validation via `base.extend(overlayShape)`.

File: addons/adapter-ui/cap-adapter-ui/src/components/auto-form/auto-form.tsx [hunk 11]
```diff
@@ -77,6 +83,17 @@ export function AutoForm({
...
+    if (overlayFields && overlayFields.length > 0) {
+      const extensions = (data?._extensions ?? {}) as Record<string, unknown>;
+      for (const overlay of overlayFields) {
+        const key = overlayFieldKey(overlay.fieldName);
+        initial[key] =
+          extensions[overlay.fieldName] ??
+          overlay.config.defaultValue ??
+          getOverlayDefault(overlay.fieldType);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (laofahai) [reply addons/adapter-ui/cap-adapter-ui/src/components/auto-form/auto-form.tsx:107]: Fixed. `handleClearTemplate()` now restores overlay field defaults using the same logic as initial state: reads from existing `_extensions` data, falls back to `overlay.config.defaultValue`, then `getOverlayDefault()`.

File: addons/adapter-ui/cap-adapter-ui/src/components/auto-form/auto-form.tsx [hunk 12]
```diff
@@ -425,6 +442,21 @@ export function AutoForm({
...
+      for (const overlay of overlayFields) {
+        const key = overlayFieldKey(overlay.fieldName);
+        if (key in submitData) {
+          extensions[overlay.fieldName] = submitData[key];
+          delete submitData[key];
+        }
+      }
+      if (Object.keys(extensions).length > 0) {
+        submitData._extensions = extensions;
+      }
```
[the remark this answers is not in the document]
Comment (laofahai) [reply addons/adapter-ui/cap-adapter-ui/src/components/auto-form/auto-form.tsx:481]: Already fixed. The code spreads existing `data?._extensions` before overlaying new values: `{ ...((data?._extensions as Record<string, unknown>) ?? {}) }`. Non-overlay extension keys are preserved.

File: addons/adapter-ui/cap-adapter-ui/src/components/auto-list/columns.tsx [hunk 13]
```diff
@@ -136,6 +143,58 @@ export function buildColumns(opts: BuildColumnsOptions): ColumnDef<DataRow>[] {
...
+              )}
+            </button>
+          );
+        },
+        cell: ({ getValue }) => {
+          const value = getValue();
+          return formatOverlayCellValue(value, overlay.fieldType);
+        },
+        enableSorting: sortable,
+        size: 150,
```
[the remark this answers is not in the document]
Comment (laofahai) [reply addons/adapter-ui/cap-adapter-ui/src/components/auto-list/columns.tsx:194]: Fixed. The `accessorFn` already reads from `row._extensions[fieldName]` correctly. Added `sortUndefined: -1` to push null/undefined overlay values to the bottom during sorting.

File: packages/core/src/overlay/promote.ts [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+
+  // Build column definition
+  let columnDef = `"${fieldName}" ${sqlType}`;
+  if (config.required) {
+    // Use a default for the NOT NULL constraint during backfill
+    const defaultExpr = getDefaultExpression(fieldType, config);
+    if (defaultExpr !== null) {
+      columnDef += ` NOT NULL DEFAULT ${defaultExpr}`;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (laofahai) [reply packages/core/src/overlay/promote.ts:139]: Fixed. Added COALESCE with getDefaultExpression fallback for all field types including date. Required fields now always get NOT NULL + DEFAULT.

File: packages/cli/__tests__/mcp-dev.test.ts [hunk 15]
```diff
@@ -99,14 +99,14 @@ async function callTool(
   args: Record<string, unknown> = {},
 ): Promise<{ content: Array<{ type: string; text: string }>; isError?: boolean }> {
   // biome-ignore lint/suspicious/noExplicitAny: accessing internal for testing
-  const registeredTools = (server as any)._registeredTools as Record<string, { handler: Function }>;
+  const registeredTools = (server as any)._registeredTools as Record<string, { handler: (...args: never[]) => unknown }>;
```
[the remark this answers is not in the document]
Comment (laofahai) [reply packages/cli/__tests__/mcp-dev.test.ts]: Fixed. Replaced `as any` with `RegisteredTool` interface + `ToolsMap` type. Extracted `getTools()` helper that casts via `unknown` instead of `any`.

File: packages/cli/__tests__/mcp-dev.test.ts [hunk 16]
```diff
@@ -121,7 +121,7 @@ describe("MCP Dev Server", () => {
   describe("linchkit_list_entities", () => {
     test("returns all entities with name, label, fieldCount", async () => {
       const result = await callTool(server, "linchkit_list_entities");
-      const data = JSON.parse(result.content[0]!.text);
+      const data = JSON.parse(result.content[0]?.text);
```
[the remark this answers is not in the document]
Comment (laofahai) [reply packages/cli/__tests__/mcp-dev.test.ts]: Fixed. Added `parseToolResult()` helper that checks for undefined before calling JSON.parse, throwing a clear error if content is missing. All 15 occurrences replaced.

File: packages/cli/src/commands/overlay.ts [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+
+        // Mark as promoted in the database
+        await conn.store.updateOverlay(overlay.id, { status: "promoted" });
```
[the remark this answers is not in the document]
Comment (laofahai) [reply packages/cli/src/commands/overlay.ts]: Fixed. Added try/catch around updateOverlay — if it fails after writing the migration file, the file is cleaned up via unlinkSync. Also added pre-existence check before writing.
